In [1]:
import pandas as pd

In [2]:
df_train_purch = pd.read_csv('https://www.dropbox.com/s/fxm6ffz3a5seu6k/train_purch.csv?dl=1')
df_clients2 = pd.read_csv('https://www.dropbox.com/s/b020sd2fdzq936p/clients2.csv?dl=1')
df_products = pd.read_csv('https://www.dropbox.com/s/1u8pz9rwk80wmfv/products.csv?dl=1')
df_train = pd.read_csv('https://www.dropbox.com/s/zgfmxpas92pkuc2/train.csv?dl=1')

### Посмотрим на данные

In [3]:
print('---TRAIN_PURCH---')
display(df_train_purch.head())
print(df_train_purch.info())
print('\n---CLIENTS2---')
display(df_clients2.head())
print(df_clients2.info())
print('\n---PRODUCTS---')
display(df_products.head())
print(df_products.info())
print('\n---TRAIN---')
display(df_train.head())
print(df_train.info())

---TRAIN_PURCH---


client_id transaction_id transaction_datetime  regular_points_received  \
0  000012768d     7e3e2e3984  2018-12-01 07:12:45                     10.0   
1  000012768d     7e3e2e3984  2018-12-01 07:12:45                     10.0   
2  000012768d     7e3e2e3984  2018-12-01 07:12:45                     10.0   
3  000012768d     7e3e2e3984  2018-12-01 07:12:45                     10.0   
4  000012768d     7e3e2e3984  2018-12-01 07:12:45                     10.0   

   express_points_received  regular_points_spent  express_points_spent  \
0                      0.0                   0.0                   0.0   
1                      0.0                   0.0                   0.0   
2                      0.0                   0.0                   0.0   
3                      0.0                   0.0                   0.0   
4                      0.0                   0.0                   0.0   

   purchase_sum    store_id  product_id  product_quantity  trn_sum_from_iss  \
0        1007.0  54a4a11a29  9a80204f78               2.0              80.0   
1        1007.0  54a4a11a29  da89ebd374               1.0              65.0   
2        1007.0  54a4a11a29  0a95e1151d               1.0              24.0   
3        1007.0  54a4a11a29  4055b15e4a               2.0              50.0   
4        1007.0  54a4a11a29  a685f1916b               1.0              22.0   

   trn_sum_from_red  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15998952 entries, 0 to 15998951
Data columns (total 13 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   client_id                object 
 1   transaction_id           object 
 2   transaction_datetime     object 
 3   regular_points_received  float64
 4   express_points_received  float64
 5   regular_points_spent     float64
 6   express_points_spent     float64
 7   purchase_sum             float64
 8   store_id                 object 
 9   product_id               object 
 10  product_quantity         float64
 11  trn_sum_from_iss         float64
 12  trn_sum_from_red         float64
dtypes: float64(8), object(5)
memory usage: 1.5+ GB
None

---CLIENTS2---


client_id  client_id.1     first_issue_date    first_redeem_date  age  \
0  000012768d          NaN  2017-08-05 15:40:48  2018-01-04 19:30:07   45   
1  000036f903          NaN  2017-04-10 13:54:23  2017-04-23 12:37:56   72   
2  00010925a5          NaN  2018-07-24 16:21:29  2018-09-14 16:12:49   83   
3  0001f552b0          NaN  2017-06-30 19:20:38  2018-08-28 12:59:45   33   
4  00020e7b18          NaN  2017-11-27 11:41:45  2018-01-10 17:50:05   73   

  gender  
0      U  
1      F  
2      U  
3      F  
4      U

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200039 entries, 0 to 200038
Data columns (total 6 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   client_id          200039 non-null  object 
 1   client_id.1        0 non-null       float64
 2   first_issue_date   200039 non-null  object 
 3   first_redeem_date  182493 non-null  object 
 4   age                200039 non-null  int64  
 5   gender             200039 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 9.2+ MB
None

---PRODUCTS---


product_id     level_1     level_2     level_3     level_4  segment_id  \
0  0003020d3c  c3d3a8e8c6  c2a3ea8d5e  b7cda0ec0c  6376f2a852       123.0   
1  0003870676  e344ab2e71  52f13dac0c  d3cfe81323  6dc544533f       105.0   
2  0003ceaf69  c3d3a8e8c6  f2333c90fb  419bc5b424  f6148afbc0       271.0   
3  000701e093  ec62ce61e3  4202626fcb  88a515c084  48cf3d488f       172.0   
4  0007149564  e344ab2e71  52f13dac0c  d3cfe81323  6dc544533f       105.0   

     brand_id   vendor_id  netto  is_own_trademark  is_alcohol  
0  394a54a7c1  9eaff48661  0.400                 0           0  
1  acd3dd483f  10486c3cf0  0.680                 0           0  
2  f597581079  764e660dda  0.500                 0           0  
3  54a90fe769  03c2d70bad  0.112                 0           0  
4  63417fe1f3  f329130198  0.600                 0           0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43038 entries, 0 to 43037
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   product_id        43038 non-null  object 
 1   level_1           43035 non-null  object 
 2   level_2           43035 non-null  object 
 3   level_3           43035 non-null  object 
 4   level_4           43035 non-null  object 
 5   segment_id        41466 non-null  float64
 6   brand_id          37838 non-null  object 
 7   vendor_id         43004 non-null  object 
 8   netto             43035 non-null  float64
 9   is_own_trademark  43038 non-null  int64  
 10  is_alcohol        43038 non-null  int64  
dtypes: float64(2), int64(2), object(7)
memory usage: 3.6+ MB
None

---TRAIN---


client_id  treatment_flg  target
0  ad6561e2d8              1       1
1  7c1ccbf93f              1       1
2  b58fadcab6              1       1
3  e99e6fabb9              0       0
4  27fb6f8520              1       1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140027 entries, 0 to 140026
Data columns (total 3 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   client_id      140027 non-null  object
 1   treatment_flg  140027 non-null  int64 
 2   target         140027 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.2+ MB
None


### Оставим только те колонки, которые нам могут понадобиться

In [4]:
df_train_purch = df_train_purch[['client_id','transaction_id','transaction_datetime','purchase_sum','product_id','product_quantity']]
df_train_purch['transaction_datetime'] = df_train_purch['transaction_datetime'].astype('datetime64')
df_clients2 = df_clients2.drop(columns = 'client_id.1')
df_clients2['first_issue_date'] = df_clients2['first_issue_date'].astype('datetime64')
df_clients2['first_redeem_date'] = df_clients2['first_redeem_date'].astype('datetime64')
df_products = df_products[['product_id','is_own_trademark']]
print('---TRAIN_PURCH---')
print(df_train_purch.info())
print('\n---CLIENTS2---')
print(df_clients2.info())
print('\n---PRODUCTS---')
print(df_products.info())

---TRAIN_PURCH---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15998952 entries, 0 to 15998951
Data columns (total 6 columns):
 #   Column                Dtype         
---  ------                -----         
 0   client_id             object        
 1   transaction_id        object        
 2   transaction_datetime  datetime64[ns]
 3   purchase_sum          float64       
 4   product_id            object        
 5   product_quantity      float64       
dtypes: datetime64[ns](1), float64(2), object(3)
memory usage: 732.4+ MB
None

---CLIENTS2---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200039 entries, 0 to 200038
Data columns (total 5 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   client_id          200039 non-null  object        
 1   first_issue_date   200039 non-null  datetime64[ns]
 2   first_redeem_date  182493 non-null  datetime64[ns]
 3   age                200039 non-null  int64  

# Вопросы в тестовом задании

In [5]:
uplift = df_train.loc[df_train['treatment_flg']==1, 'target'].mean() - df_train.loc[df_train['treatment_flg']==0, 'target'].mean()
train_list = list(df_train['client_id'])

up_50 = df_clients2[(df_clients2['client_id'].isin(train_list)) & (df_clients2['age']>=50)]['client_id']
down_50 = df_clients2[(df_clients2['client_id'].isin(train_list)) & (df_clients2['age']<50)]['client_id']
uplift_50up = df_train.loc[(df_train['treatment_flg']==1 & df_train['client_id'].isin(up_50)), 'target'].mean() - df_train.loc[(df_train['treatment_flg']==0 & df_train['client_id'].isin(up_50)), 'target'].mean()
uplift_50down = df_train.loc[(df_train['treatment_flg']==1 & df_train['client_id'].isin(down_50)), 'target'].mean() - df_train.loc[(df_train['treatment_flg']==0 & df_train['client_id'].isin(down_50)), 'target'].mean()

df_train_timegaps = df_train_purch[['client_id','transaction_datetime']].drop_duplicates().groupby('client_id').apply(lambda x: x.diff().mean())

own_product_list = list(df_products[df_products['is_own_trademark']==1]['product_id'])

### Вопрос 1

Какое значение аплифта на всем трейне (train.csv)

In [6]:
print(f"Аплифт на всем трейне {uplift:.2}\n")

Аплифт на всем трейне 0.034



### Вопрос 2

Выберите правильные утверждения:  
1) В трейне мужчин больше чем женщин  
2) Медианный возраст всех людей в трейне равен 39 годам  
3) Последняя покупка в train_purch была совершена в марте 2018 года  
4) Аплифт для людей старше 50 лет больше чем в два раза выше чем у остальных людей

In [7]:
print(f"1. Возрастное распределение клиентов в трейне\n{df_clients2[df_clients2['client_id'].isin(train_list)].groupby('gender')['client_id'].count().sort_values(ascending=False)}\nЖеншин больше, чем мужчин\n")
print(f"2. Медианный возраст всех людей в трейне {df_clients2[df_clients2['client_id'].isin(train_list)]['age'].median()}\n")
print(f"3. Последняя покупка в трейне состоялась {df_train_purch['transaction_datetime'].max()}\n")
print(f"4. Аплифт 50+ в {uplift_50up / uplift_50down:.2} раза больше аплифта остальных")

1. Возрастное распределение клиентов в трейне
gender
U    64965
F    51559
M    23503
Name: client_id, dtype: int64
Женшин больше, чем мужчин

2. Медианный возраст всех людей в трейне 45.0

3. Последняя покупка в трейне состоялась 2019-03-18 23:19:28

4. Аплифт 50+ в 1.8 раза больше аплифта остальных


### Вопрос 3

На какую максимальную сумму в train_purch была совершена покупка? (округлить до целого)  
- 21983  
- 29611  
- 44259  
- 54364  
- Здесь такого числа нет

In [8]:
print(f"Максимальная сумма покупки {df_train_purch['purchase_sum'].max()}")

Максимальная сумма покупки 29611.48


### Вопрос 4

Раз в сколько дней в среднем покупатели ходят за покупками? (округлить до целого)

In [9]:
print(f"Средняя частота покупок {df_train_timegaps['transaction_datetime'].mean()}")

Средняя частота покупок 9 days 07:32:26.770712154


Варианты ответа 2, 3, 4, 5. У меня получилось 9. Непонятно.

### Вопрос 5

Какого значения аплифта можно достичь, если осуществить воздействие на людей, с момента последней покупки которых прошло 10 и более дней?  
- 3.02%  
- 3.72%  
- 4.15%  
- 6.32%

Не понимаю. У меня есть даты покупок, но нет дат воздействия на клиентов. Как сортировать?  
Предположил, что время воздействия как-то связано с first_issue_date во фрейме clients2. Но максимальная дельта там 15 часов. Ни о каких 10 днях и более речь не идет.

### Вопрос 6

Какова доля продукции собственного производства в покупках среди всех покупок?  
- 3.42%  
- 11.2%  
- 13.1%  
- 15.2%

In [10]:
print(f"Доля собственных продуктов в продажах составляет {df_train_purch[df_train_purch['product_id'].isin(own_product_list)]['transaction_id'].count() / df_train_purch['transaction_id'].count():.4}")

Доля собственных продуктов в продажах составляет 0.1523


### Вопрос 7

Влияет ли факт покупки самого популярного продукта на аплифт? Если влияет, то каково значение аплифта?  
1. Не влияет  
2. Влияет, 0.35%  
3. Влияет, 2.64%  
4. Влияет, 4.86%

In [11]:
# Самый популярный продукт
pop_product_id = df_train_purch['product_id'].value_counts().index[0]
# список клиентов, купивших этот продукт
pop_prod_client_list = list(df_train_purch[df_train_purch['product_id']==pop_product_id]['client_id'].drop_duplicates())
uplift_pop = df_train.loc[(df_train['client_id'].isin(pop_prod_client_list) & df_train['treatment_flg']==1), 'target'].mean() - df_train.loc[(df_train['client_id'].isin(pop_prod_client_list) & df_train['treatment_flg']==0), 'target'].mean()
uplift_pop / uplift

0.9826499602476436

Похоже, что аплифт упал на 1.7%

In [12]:
print(f"Значение аплифта с самым популярным товаром {uplift_pop:.2}")

Значение аплифта с самым популярным товаром 0.033


### Вопрос 8

Разделим покупателей на две группы: владеют картой покупателя > 100 дней (группа А), владеют картой покупателя <= 100 дней (группа Б). Правда ли то, что аплифт в группе А меньше, чем в группе Б? Если правда, то на сколько меньше (разница между процентами)?  
- Не правда  
- Правда, 1%  
- Правда, 2%  
- Правда, 3%

Не понимаю, где взять срок владения.

### Вопрос 9

Разделим покупателей на две группы: покупали продукты собственного производста хоть раз (группа А), вообще не покупали продукты собственного производства (группа Б). Правда ли то, что аплифт в группе А меньше чем в группе Б? Если правда, то на сколько меньше?  
- Не правда  
- Правда, всего 1%  
- Правда, 2%  
- Правда, почти 5%  

In [13]:
# список продуктов собственного производства
own_prod_list = list(df_products[df_products['is_own_trademark']==1]['product_id'].drop_duplicates())
# список всех покупателей
cl_list = set(df_train_purch['client_id'])
# список покупателей собственной продукции
own_cl_list = set(df_train_purch[df_train_purch['product_id'].isin(own_prod_list)]['client_id'])
non_own_cl_list = list(cl_list - own_cl_list)
own_cl_list = list(own_cl_list)

uplift_own = df_train.loc[(df_train['client_id'].isin(own_cl_list) & df_train['treatment_flg']==1), 'target'].mean() - df_train.loc[(df_train['client_id'].isin(own_cl_list) & df_train['treatment_flg']==0), 'target'].mean()
uplift_non_own = df_train.loc[(df_train['client_id'].isin(non_own_cl_list) & df_train['treatment_flg']==1), 'target'].mean() - df_train.loc[(df_train['client_id'].isin(non_own_cl_list) & df_train['treatment_flg']==0), 'target'].mean()
print(f"Аплифт клиентов, которые покупали продукцию нашего производства: {uplift_own}")
print(f"Аплифт клиентов, которые никогда не покупали продукцию нашего производства: {uplift_non_own}")

Аплифт клиентов, которые покупали продукцию нашего производства: 0.0467519621487984
Аплифт клиентов, которые никогда не покупали продукцию нашего производства: -0.22136412819892948


### Вопрос 10

Разделим покупателей на две группы: сумма покупок которых больше среднего чека по всей выборке (группа А), сумма покупок которых меньше среднего чека по всей выборке (группа Б). Правда ли то, что аплифт в группе А меньше, чем в группе Б?  
- Не правда
- Правда, 0.6%
- Правда, 1.3%
- Правда, 2.1%

In [14]:
client_spendings = df_train_purch[['client_id','purchase_sum']].groupby('client_id').sum().reset_index()
mean_sum = client_spendings['purchase_sum'].mean()
client_spendings_a = list(client_spendings[client_spendings['purchase_sum'] > mean_sum]['client_id'])
client_spendings_b = list(client_spendings[client_spendings['purchase_sum'] < mean_sum]['client_id'])
uplift_a = df_train.loc[(df_train['client_id'].isin(client_spendings_a) & df_train['treatment_flg']==1), 'target'].mean() - df_train.loc[(df_train['client_id'].isin(client_spendings_a) & df_train['treatment_flg']==0), 'target'].mean()
uplift_b = df_train.loc[(df_train['client_id'].isin(client_spendings_b) & df_train['treatment_flg']==1), 'target'].mean() - df_train.loc[(df_train['client_id'].isin(client_spendings_b) & df_train['treatment_flg']==0), 'target'].mean()
print(f"Аплифт в группе А {uplift_a}")
print(f"Аплифт в группе B {uplift_b}")

Аплифт в группе А 0.14612674040560936
Аплифт в группе B -0.038542523836676


### Предположение по 5 вопросу

In [15]:
df_train_purch_date = df_train_purch[['client_id','transaction_id','transaction_datetime']].drop_duplicates()
df_clients_cut = df_clients2[['client_id','first_issue_date']]
df_dates = df_train_purch_date.merge(df_clients_cut, on='client_id')
df_dates2 = df_dates[df_dates['first_issue_date'] > df_dates['transaction_datetime']]
df_dates2['delta'] = df_dates2['first_issue_date'] - df_dates2['transaction_datetime']
print(df_dates2.delta.mean())
df_dates2.sort_values('delta',ascending=False)

0 days 02:54:55.461150667


<ipython-input-15-41b16ba3d09f>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dates2['delta'] = df_dates2['first_issue_date'] - df_dates2['transaction_datetime']


client_id transaction_id transaction_datetime    first_issue_date  \
222373   149478b823     6ed4148a00  2018-12-31 12:50:25 2019-01-01 04:05:52   
1298198  7689a9ea9d     3841d17f7f  2019-02-26 11:09:17 2019-02-26 20:31:17   
165314   0f6cd42f6a     ed732dfb5b  2019-02-26 10:46:01 2019-02-26 19:25:04   
1725975  9d47bdc407     d54af4e399  2018-12-11 12:25:43 2018-12-11 20:08:26   
2525247  e5e14ce6d4     7b48b47a70  2018-11-29 09:15:32 2018-11-29 16:50:29   
...             ...            ...                  ...                 ...   
387254   237f75c059     043f06a8ab  2018-12-30 10:47:14 2018-12-30 10:48:35   
1831357  a7021ad220     b1ceaa2b37  2018-12-28 14:06:02 2018-12-28 14:07:13   
2050275  babdc21558     3874cb28e0  2019-03-08 10:22:31 2019-03-08 10:23:03   
198597   127163689c     730a0d5ec0  2018-12-19 17:04:02 2018-12-19 17:04:19   
387255   237f75c059     e5d7756a27  2018-12-30 10:48:24 2018-12-30 10:48:35   

                  delta  
222373  0 days 15:15:27  
1298198 0 days 09:22:00  
165314  0 days 08:39:03  
1725975 0 days 07:42:43  
2525247 0 days 07:34:57  
...                 ...  
387254  0 days 00:01:21  
1831357 0 days 00:01:11  
2050275 0 days 00:00:32  
198597  0 days 00:00:17  
387255  0 days 00:00:11  

[19241 rows x 5 columns]